src/compiler/moduleNameResolver.ts

```diff
@@ -402,8 +402,10 @@ namespace ts {
      */
     export function getAutomaticTypeDirectiveNames(options: CompilerOptions, host: ModuleResolutionHost): string[] {
         // Use explicit type list from tsconfig.json
+        // jsxImportSource, if present and in use, creates implicit imports
+        const implicitImport = getJSXRuntimeImport(getJSXImplicitImportBase(options), options);

```

Comment:
I am not sure we should be adding anything implicit if `types` is specified.  Thats exactly why that option exists.

Reply:
I add it to the `types` list because I don't think there's a great reason to make people repeat the same (or nearly the same) module name for both the `jsxImportSource` and the `types` flags - it just seems to me to make sense to include it in the `types` list, if present.

Reply:
Question for @DanielRosenwasser then. If we are always going to implicitly add this we should process this separately in processTypeReferenceDirective in program.ts instead of modifying list here. 

I think apart from incremental we should add test where the file is added and later goes missing for this implicit import with and without types specified because we handle them specifically for watching in resolutionCache

Reply:
Isn't this sort of a similar concept to how `tslib` is added with `importHelpers`? You've got an implicit import injected and we do some resolution to see if it's vaguely compatible, right?

@RyanCavanaugh and @andrewbranch might want to weigh in here as well.

Reply:
Yeah, I guess so - primary difference being we actually end up pulling the types from the module for checking things.

Reply:
Looks like we might do the same for tslib?

```ts
function checkExternalEmitHelpers(location: Node, helpers: ExternalEmitHelpers) {
    if ((requestedExternalEmitHelpers & helpers) !== helpers && compilerOptions.importHelpers) {
        const sourceFile = getSourceFileOfNode(location);
        if (isEffectiveExternalModule(sourceFile, compilerOptions) && !(location.flags & NodeFlags.Ambient)) {
            const helpersModule = resolveHelpersModule(sourceFile, location);
            if (helpersModule !== unknownSymbol) {
                const uncheckedHelpers = helpers & ~requestedExternalEmitHelpers;
                for (let helper = ExternalEmitHelpers.FirstEmitHelper; helper <= ExternalEmitHelpers.LastEmitHelper; helper <<= 1) {
                    if (uncheckedHelpers & helper) {
                        const name = getHelperName(helper);
                        const symbol = getSymbol(helpersModule.exports!, escapeLeadingUnderscores(name), SymbolFlags.Value);
                        if (!symbol) {
                            error(location, Diagnostics.This_syntax_requires_an_imported_helper_named_1_which_does_not_exist_in_0_Consider_upgrading_your_version_of_0, externalHelpersModuleNameText, name);
                        }
                    }
                }
            }
            requestedExternalEmitHelpers |= helpers;
        }
    }
}
```

Reply:
Hm, for `importHelpers` we make a synthetic `import` within `collectExternalModuleReferences` - I could do that; hopefully it doesn't have other effects?

Reply:
Not sure - @rbuckton might know more here too.

Reply:
I've changed the implementation to use the same synthetic-import method that we use for `importHelpers`.